File: 06-Code/RolePaymentSystem_V2.0_Python/utils/Calculator.py
from model.Constants import Constants
class Calculator:
    @staticmethod
    def calculate_basic_salary(annual_salary):
        return annual_salary / 12

    @staticmethod
    def calculate_overtime_hours(hours_worked, regular_hours, hourly_rate):
        overtime_hours = hours_worked - regular_hours
        overtime_value = overtime_hours * hourly_rate * Constants.OVERTIME_HOUR_INCREASE
        return overtime_value

    @staticmethod
    def calculate_reserve_funds(basic_salary):
        return basic_salary * Constants.RESERVE_FUNDS_PERCENTAGE

    @staticmethod
    def calculate_iess_contribution(total_income, reserve_funds):
        iess_contribution = (total_income - reserve_funds) * Constants.IESS_PERCENTAGE
        return iess_contribution

    @staticmethod
    def calculate_biweekly_advance(basic_salary):
        return basic_salary / 2

    @staticmethod
    def calculate_food_deduction(brings_own_food, food_deduction_amount):
        return 0 if brings_own_food else food_deduction_amount

    @staticmethod
    def calculate_total_income(basic_salary, overtime_payment, bonuses):
        return basic_salary + overtime_payment + bonuses

    @staticmethod
    def calculate_total_expenses(iess_contribution, biweekly_advance, iess_loans, company_loans, fines, food_deduction):
        return iess_contribution + biweekly_advance + iess_loans + company_loans + fines + food_deduction

    @staticmethod
    def calculate_net_payment(total_income, total_expenses):
        return total_income - total_expenses

    @staticmethod
    def calculate_employer_contribution(total_income, reserve_funds):
        return (total_income + reserve_funds) * Constants.EMPLOYER_CONTRIBUTION_PERCENTAGE

    @staticmethod
    def calculate_total_employee_cost(net_payment, employer_contribution):
        return net_payment + employer_contribution

File: 06-Code/GraphicInterface/src/main/java/ec/espe/edu/rolepaymentsystem/util/Calculator.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package ec.espe.edu.rolepaymentsystem.util;

import ec.espe.edu.rolepaymentsystem.model.Constants;
import ec.espe.edu.rolepaymentsystem.model.Employee;

/**
 *
 * @author PAOLA-SSD
 */
public class Calculator {
    public double calculateTotalAmount(Employee emp) {
    double totalAmount = emp.getBasicSalary();
    totalAmount += emp.getBonuses();
    totalAmount += emp.getOvertimeHours() * (emp.getBasicSalary() / 160);
    totalAmount -= emp.getIessLoans();
    totalAmount -= emp.getCompanyLoans();
    totalAmount -= emp.getFines();
    double dailySalary = emp.getBasicSalary() / 30; 
    totalAmount -= emp.getAbsentDays() * dailySalary;
    
    return Math.max(totalAmount, 0);
}
    public static double OtherDeduccion(double totalDeductions, Employee emp) {
        return totalDeductions += emp.getAbsentDays() + emp.getFines() + emp.getCompanyLoans() + emp.getIessLoans();
    }
    public static double calculateOvertimeHours(double hoursWorked, double regularHours, double hourlyRate) {
        double overtimeHours = hoursWorked - regularHours;
        double overtimeValue = overtimeHours * hourlyRate * Constants.overtimeHourIncrease;
        return overtimeValue;
    }

    public static double calculateReserveFunds(double basicSalary) {
        return basicSalary * Constants.reserveFundsPercentage;
    }

    public static double calculateIessContribution(double totalIncome, double reserveFunds) {
        double iessContribution = (totalIncome - reserveFunds) * Constants.iessPercentage;
        return iessContribution;
    }
    
    public static double calculateBiweeklyAdvance(double basicSalary) {
    return basicSalary / 2;
    }
    
    public static double calculateFoodDeduction(boolean bringsOwnFood, double foodDeductionAmount) {
    return bringsOwnFood ? 0 : foodDeductionAmount;
    }

    public static double calculateTotalIncome(double basicSalary, double overtimePayment, double bonuses) {
        return basicSalary + overtimePayment + bonuses;
    }

    public static double calculateTotalExpenses(double iessContribution, double biweeklyAdvance, double iessLoans, double companyLoans, double fines, double foodDeduction) {
        return iessContribution + biweeklyAdvance + iessLoans + companyLoans + fines + foodDeduction;
    }

    public static double calculateNetPayment(double totalIncome, double totalExpenses) {
        return totalIncome - totalExpenses;
    }

    public static double calculateEmployerContribution(double totalIncome, double reserveFunds) {
        return (totalIncome + reserveFunds) * Constants.employerContributionPercentage;
    }

    public static double calculateTotalEmployeeCost(double netPayment, double employerContribution) {
        return netPayment + employerContribution;
    }
}
